lib/spell-check-manager.coffee

```diff
@@ -0,0 +1,306 @@
+class SpellCheckerManager
+  checkers: []
+  locales: []
+  localePaths: []
+  useLocales: false
+  localeCheckers: null
+  knownWords: []
+  addKnownWords: false
+  knownWordsChecker: null
+
+  addPluginChecker: (checker) ->
+    console.log "spell-check: addPluginChecker:", checker
+    @addSpellChecker checker
+  addSpellChecker: (checker) ->
+    console.log "spell-check: addSpellChecker:", checker
+    @checkers.push checker
+
+  removeSpellChecker: (spellChecker) ->
+    @checkers = @checkers.filter (plugin) -> plugin isnt spellChecker
+
+  check: (args, text) ->
+    # Make sure our deferred initialization is done.
+    @init()
+
+    # We need a couple packages.
+    multirange = require 'multi-integer-range'
```

Comment:
> In the worst case, checking every word on 100k words is about double what it is to send the entire buffer as a single string.

If we manage to wrap this inside a `Task`, I feel like this might already be fine. Moreover, we can let the UI thread be updated at regular intervals, so that it doesn't look like no spell-checking is being performed in the case of a long text. :racehorse: 

Also, I am wondering if you could setup a test where, instead of crossing the native boundary **for each word**, we pass an array of strings (the words), and return an array of booleans. It seems like it could greatly improve performance because I'd assume `strings` get passed by reference, and `bool` is a really cheap data type. What do you think?

> think with a small cache (5k words, settable via config), it would be advantageous to switch to a word-by-word. It would let us drop the dependency on multi-integer-range and simplify the logic for marking up the errors.

I think it'd be nice to avoid the cache, just so we can keep the memory footprint to a minimum. Also, it seems like we would need to invalidate this cache when e.g. new packages that provide spell-checking get activated, and I'd :heart: to avoid this extra complexity if we can.

> do you want it as a delta against this one or rebase it off the head as a single commit? (I'd squash everything before asking you for the final pull.)

Personally, I think it's fine to avoid rebasing: feel free to just add new commits on top of the ones you've already pushed, so that this conversation can be read in the future within its context.